### ISSUE TITLE
Downloading the Model to my drive FAILED

### ISSUE THREAD

**Issue by pumpkinband** [re-search/DocProduct#4]
`!pip install -U -q PyDrive
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials

# 1. Authenticate and create the PyDrive client.
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)

# 2. Save Keras Model or weights on google drive

# create on Colab directory
model.save('model.h5')    
model_file = drive.CreateFile({'title' : 'model.h5'})
model_file.SetContentFile('model.h5')
model_file.Upload()

# download to google drive
drive.CreateFile({'id': model_file.get('id')})`

Used this code to Download the model but Failed
 
ERROR :- 

`NameError: name 'model' is not defined`

**Comment by Santosh-Gupta** [comment]
It looks like you used the train your own models notebook, and now you're trying to save the model you trained? Or is it something else?

**Comment by pumpkinband** [comment]
yes sir i want to download model.h5 , can you help me please.
@Santosh-Gupta

**Comment by JayYip** [comment]
What are you trying to download? The model weights trained by yourself or the weights that we provide?

**Comment by pumpkinband** [comment]
Model weight i trained i guess. (i meant running the CELLS by my self) @JayYip

**Comment by JayYip** [comment]
May I know which notebook are you trying to run?

**Comment by pumpkinband** [comment]
https://colab.research.google.com/drive/1FSti89LtRIzuQapRpwv2Lih4Dm3dENHb#scrollTo=qZ110ZqNfBac   @JayYip

**Comment by JayYip** [comment]
Seems I cannot open the link you provide. That might be a copy of the original colab notebook and I cannot access it. Which notebook in this [folder](https://drive.google.com/drive/folders/1kXqsE4N0MgfktsEJpZZn470yOZ3UQ10F)?

**Comment by pumpkinband** [comment]
DocProductPresentationV6-0.2.0.ipynb

**Comment by JayYip** [comment]
Sorry, but I can't find the corresponding code in the notebook.

**Comment by pumpkinband** [comment]
Sure sir but how can i download model.h5 of your model from OneDrive, can you please direct me.

**Comment by JayYip** [comment]
I'm not sure what is `model.h5`. As for the model weights that we provide, you can download [here](https://onedrive.live.com/download?cid=9DEDF3C1E2D7E77F&resid=9DEDF3C1E2D7E77F%2132792&authkey=AEQ8GtkcDbe3K98).

**Comment by pumpkinband** [comment]
ok but how to download the model weight can you please help me step by step

**Comment by JayYip** [comment]
The zip file that I linked above contains the model weights. That is the same file as `DataAndCheckpoint.zip` in DocProductPresentationV6-0.2.0.ipynb notebook. 

Then you need to upload the file you downloaded to colab and rename it to `DataAndCheckpoint.zip` and execute the notebook. 

I am still confused though, what is the `model` in your original post? Where do you get that chunk of code?

**Comment by harshit-saraswat** [comment]
Hello sir, actually I had an issue with the model. I've downloaded the weights and embedding file from your  drive and when i tried to run it on my machine(windows os) i got an error in predictor.py file as it imports faiss which isn't available on windows yet. So, I was thinking is there a way to run the model on windows? Also I'm trying to deploy the model on an android based application and the video Siraj posted on his channel doesn't actually covers it neither his repository. Mind shedding in some light?

**Comment by pumpkinband** [comment]
same issue as @harshit-saraswat

**Comment by JayYip** [comment]
> So, I was thinking is there a way to run the model on windows?

Please take a look at [this issue](https://github.com/facebookresearch/faiss/issues/334). Seems they're working on a Windows version.

> Also I'm trying to deploy the model on an android based application and the video Siraj posted on his channel doesn't actually covers it neither his repository

I don't have many experience of developing apps but as far as I can tell, retrieval based model is not very suitable to put the whole model to mobile device because the embedding vector size is huge. I would suggest you to serve the model on a server.

BTW, if you guys have experience developing apps, maybe you can work with us! If you're interested please contact me via email: junpeng.yip@gmail.com and let us get to know you and your skillset.

**Comment by harshit-saraswat** [comment]
@JayYip there's the problem right, for the model to be used as a server I do need it to be run on my machine as for as far as I know, colab doesn't allow webhooks or post requests from other consoles/webservices such as dialogflow. TBH integrating it with Dialogflow as a chatbot would really be a great deal. I'm really trying on howto send my inputted query string to your model online as an server and get the ouput as result. Any ideas about how can I use this functionality? It needn't be necessary to be used as a dialogflow integration/ backend. I just need some guidance about using the model as a server and sending request and retrieving the output back, since as of now colab doesn't serve that type of services/purposes. Thanks!

**Comment by pumpkinband** [comment]
on that topic i want to convert it into .tflite can you help me with that @harshit-saraswat

**Comment by harshit-saraswat** [comment]
@pumpkinband Follow the following steps:
1. Convert the model checkpoints which you've downloaded from the one drive link given by the creators here, to a h5/pb file first. These links will help- https://www.tensorflow.org/tutorials/keras/save_and_restore_models,https://stackoverflow.com/questions/44466066/how-can-i-convert-a-trained-tensorflow-model-to-keras, https://blog.metaflow.fr/tensorflow-how-to-freeze-a-model-and-serve-it-with-a-python-api-d4f3596b3adc.
2. Now use this link to convert the saved graph instances of the model to a tflite file. https://www.tensorflow.org/lite/convert/python_api

Hope it helped.

**Comment by pumpkinband** [comment]
thanks alot @harshit-saraswat and sir @JayYip thank you so much

**Comment by Santosh-Gupta** [comment]
@Jay 'model.h5' is the name of the file, what that person is doing trying to use the keras function `model.save('model.h5')` to save the weights into Keras's h.5 format. 

So the issue is that either Keras is not saving the model correctly, or that Pydrive is not uploading it correctly. 

@pumpkinband did you get your issue fixed?

**Comment by harshit-saraswat** [comment]
Anyone able to run the model?or deploy it on android yet?

**Comment by JayYip** [comment]
> what that person is doing trying to use the keras function model.save('model.h5') to save the weights into Keras's h.5 format

Yeah. I know. I'm just confused what's the value of that file since I cannot find that chunk of code in the notebook. And he did not provide any info about `model` object. 

> Anyone able to run the model?

You should be able to run the model like we did in the notebooks once you downloaded the zip file and get the environment right.

**Comment by pumpkinband** [comment]
![Screenshot (4)](https://user-images.githubusercontent.com/39296817/60318005-ef420d80-998e-11e9-847b-accc804b19b6.png)
 so can anyone tell me the name of model it is asking, i just want to download the Model in keras (h.5) ??

**Comment by harshit-saraswat** [comment]
@pumpkinband can you send a link to your code?

**Comment by pumpkinband** [comment]
@harshit-saraswat Thanks a Lot  https://colab.research.google.com/drive/19_ZL7Wr-R0ZvO-ZJkkMeg9WSGsS5rSrl#scrollTo=rVBFeL7Abhiq

**Comment by pumpkinband** [comment]
can you tell me where did i go wrong @harshit-saraswat

**Comment by harshit-saraswat** [comment]
@pumpkinband allow access for the link please. maybe change the folders sharing to allow everyone to view.

**Comment by pumpkinband** [comment]
https://colab.research.google.com/drive/19_ZL7Wr-R0ZvO-ZJkkMeg9WSGsS5rSrl  yup did it @harshit-saraswat

**Comment by JayYip** [comment]
@pumpkinband Your code doesn't make much sense to me. What do you want to download to your drive exactly? The `model` object just came out of nowhere and thus it yields NameError. If you want the model weight, it's in the path `newFolder/models/bertffn_crossentropy/bertffn`.

**Comment by harshit-saraswat** [comment]
@pumpkinband you don't need to convert the model to a json or h5 format because the model weight and graph are stored in the BioBert folder and also the new folder as .ckpt and .meta file along with some .json file. Now use those files and convert them into a .pb and .pbtxt format and then convert it into a .tflite format.

**Comment by pumpkinband** [comment]
I am trying to convert it into .pb but I just need one thing, what is output_node_names of this project ??

**Comment by Santosh-Gupta** [comment]
I need to brush up machine learning model storage types. What is the difference between how it's currently stored and .pb?  Also, what are 'output_node_names '. I'm not familiar with these, but if you can give more details I may be able to figure out what parts of DocProduct these belong in.

**Comment by pumpkinband** [comment]
3 different ways to find output_node_names, but I can’t do any of this :

1, If you still have your python file describing the model, you can always add an my_var = tf.indentity(my_var, name="my_new_name") and then load your previous model and save it again with this new added node

2, Also note that the graph_def is holding nodes in order (per connected graph), so output nodes should be close to the end of the array.

3, Finally, you can always use tensorboard to load an existing graph and just look at the name of your “output” nodes.

**Comment by pumpkinband** [comment]
can anyone please upload graph.pbtxt to tensorboard and tell me the "output_node_names" ??

**Comment by Santosh-Gupta** [comment]
What are 'output_node_names '? I'm not sure what this is. Is it the name of the output variables in the Tensorflow graph? 

"3 different ways to find output_node_names, but I can’t do any of this :"

What issues have you run into doing these?

**Comment by JayYip** [comment]
#6

**Comment by Chrome3912** [comment]
try the api: https://pocket-derm.firebaseapp.com/
